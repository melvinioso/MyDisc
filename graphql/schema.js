import 'graphql-import-node';
import { makeExecutableSchema } from 'graphql-tools';
import { applyMiddleware } from 'graphql-middleware';

import typeDefs from './typeDefs';
import resolvers from './resolvers';
import { permissions } from './permissions';
import { listArgs, defaultArgs } from './directives';

const schema = makeExecutableSchema({
  typeDefs,
  resolvers,
  schemaDirectives: {
    listArgs,
    defaultArgs,
  },
});

export default applyMiddleware(schema, permissions);
