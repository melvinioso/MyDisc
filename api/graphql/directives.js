import { GraphQLString, GraphQLInt, GraphQLNonNull } from 'graphql';
import { SchemaDirectiveVisitor } from 'graphql-tools';
import GraphQLJSON from 'graphql-type-json';

export class listArgs extends SchemaDirectiveVisitor {
  visitFieldDefinition(field) {
    field.args.push({
      name: 'order',
      type: GraphQLString,
    });
    field.args.push({
      name: 'where',
      type: GraphQLJSON,
    });
    field.args.push({
      name: 'offset',
      type: GraphQLInt,
    });
    field.args.push({
      name: 'limit',
      type: GraphQLInt,
    });
  }
}

export class defaultArgs extends SchemaDirectiveVisitor {
  visitFieldDefinition(field) {
    field.args.push({
      name: 'id',
      type: new GraphQLNonNull(GraphQLInt),
    });
    field.args.push({
      name: 'where',
      type: GraphQLJSON,
    });
  }
}
