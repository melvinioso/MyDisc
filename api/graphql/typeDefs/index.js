import { gql } from 'apollo-server-express';

import Disc from './disc';
import Bag from './bag';
import Profile from './profile';
import Email from './email';
import Permission from './permission';
import User from './user';

const Query = gql`
  scalar JSON
  scalar DateTime

  directive @listArgs on FIELD_DEFINITION
  directive @defaultArgs on FIELD_DEFINITION

  type Query {
    _empty: String
  }
`;

const Mutation = gql`
  type Mutation {
    _empty: String
  }
`;

const Subscription = gql`
  type Subscription {
    _empty: String
  }
`;

const typeDefs = [
  Query,
  Mutation,
  Subscription,
  Disc,
  Bag,
  Profile,
  Email,
  Permission,
  User,
];

export default typeDefs;
