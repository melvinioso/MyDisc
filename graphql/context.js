import db from '../models/index';

function context({ req, res }) {
  return {
    req,
    db,
  };
}

export default context;
