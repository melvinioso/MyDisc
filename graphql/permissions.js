import { rule, shield, allow, and, or } from 'graphql-shield';
import { get, includes, camelCase, merge } from 'lodash';
import pluralize from 'pluralize';
import config from '../config/config';

const hasPermission = (permission) =>
  rule(`permission.${permission}`, {
    cache: 'contextual',
  })(async (_parent, _args, ctx, _info) => {
    if (!includes(get(ctx, 'user.permissions'), permission)) {
      if (config.environment !== 'production') {
        return `Missing permission for ${permission}`;
      } else {
        return 'Not authorized.';
      }
    }
    return true;
  });

const isAuthenticated = rule({ cache: 'contextual' })(
  async (_root, _args, ctx, _info) => {
    if (ctx.user === null || ctx.user === undefined) {
      return 'Not authorized.';
    }

    return true;
  }
);

const notAllowed = rule()(async () => 'Not allowed.');

const modelKeys = [
  'Disc',
  'Bag',
  'Profile',
  'Email',
  // 'User',
  // 'Permission',
];

const defaultPermissions = modelKeys.reduce(
  (memo, modelName) => {
    const singular = camelCase(modelName);
    const plural = pluralize(singular);

    memo.Query[plural] = and(
      isAuthenticated,
      hasPermission(`${singular}.list`)
    );
    memo.Query[singular] = and(
      isAuthenticated,
      hasPermission(`${singular}.read`)
    );

    memo.Mutation[`create${modelName}`] = and(
      isAuthenticated,
      hasPermission(`${singular}.create`)
    );
    memo.Mutation[`update${modelName}`] = and(
      isAuthenticated,
      hasPermission(`${singular}.update`)
    );
    memo.Mutation[`destroy${modelName}`] = and(
      isAuthenticated,
      hasPermission(`${singular}.destroy`)
    );

    memo[modelName] = isAuthenticated;

    return memo;
  },
  { Query: {}, Mutation: {}, Subscription: {} }
);

export const permissions = shield(
  merge(defaultPermissions, {
    Query: {},
    Mutation: {},
  }),
  {
    allowExternalErrors: true,
    fallbackRule: allow,
  }
);
