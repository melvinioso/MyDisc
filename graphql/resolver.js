import { get, startCase, camelCase, omit, merge } from 'lodash';
import { resolver as _resolver } from 'graphql-sequelize';
import { EXPECTED_OPTIONS_KEY } from 'dataloader-sequelize';
import pluralize from 'pluralize';

_resolver.contextToOptions = {
  dataloaderContext: [EXPECTED_OPTIONS_KEY],
};

export function resolver(model, options = {}) {
  return async function (root, args, ctx, info) {
    model = typeof model === 'string' ? ctx.db[model] : model;
    return _resolver(model, options)(root, args, ctx, info);
  };
}

export function resolveAssociation(modelName, associationName) {
  return async function (root, args, ctx, info) {
    const associationPath = `db.${modelName}.associations.${associationName}`;
    const association = get(ctx, associationPath);

    if (!association) {
      throw new Error(
        `There is no association ${associationName} on ${modelName}`
      );
    }

    const resolved = await _resolver(association)(root, args, ctx, info);
    return resolved;
  };
}

export function buildFields(modelName, graphKeys) {
  return graphKeys.reduce((memo, item) => {
    const associationName = startCase(item).replace(/\s/gi, '');
    memo[item] = resolveAssociation(modelName, associationName);
    return memo;
  }, {});
}

export function create(modelName) {
  if (!modelName) {
    throw new Error('Please provide model name as string');
  }
  const camel = camelCase(modelName);
  return async (_, args, ctx) => {
    const record = await ctx.db[modelName].create(args[camel]);
    return record;
  };
}

export function update(modelName) {
  if (!modelName) {
    throw new Error('Please provide model name as string');
  }
  const camel = camelCase(modelName);
  return async (_, args, ctx) => {
    try {
      const attrs = get(args, camel);
      const id = get(attrs, 'id') || get(attrs, 'uuid');

      if (!id) {
        throw new Error('Missing id in args.');
      }

      const record = await ctx.db[modelName].findByPk(id);

      if (!record) {
        throw new Error(`${modelName} not found.`);
      }

      record.update(omit(attrs, ['id', 'uuid']));

      return record;
    } catch (e) {
      console.log(e);
      throw new Error(`Error updating ${modelName}.`);
    }
  };
}

export function destroy(modelName) {
  if (!modelName) {
    throw new Error('Please provide model name as string');
  }
  const camel = camelCase(modelName);
  return async (_, args, ctx) => {
    try {
      const attrs = get(args, camel);
      const id = get(attrs, 'id') || get(attrs, 'uuid');

      if (!id) {
        throw new Error('Missing id in args.');
      }

      const record = await ctx.db[modelName].findByPk(id);

      if (!record) {
        throw new Error(`${modelName} not found.`);
      }

      await record.destroy();

      return record;
    } catch (e) {
      console.log(e);
      throw new Error(`Error destroying ${modelName}`);
    }
  };
}

export function generateResolvers(modelName, fields, overrides = {}) {
  let resolvers = {
    Query: {},
    Mutation: {},
    Subscription: {},
  };

  const singular = camelCase(modelName);
  const plural = pluralize(singular);

  resolvers.Query[plural] = resolver(modelName);
  resolvers.Query[singular] = resolver(modelName);
  resolvers.Mutation[`create${modelName}`] = create(modelName);
  resolvers.Mutation[`update${modelName}`] = update(modelName);
  resolvers.Mutation[`destroy${modelName}`] = destroy(modelName);

  resolvers[modelName] = {
    ...buildFields(modelName, fields),
  };

  return merge(resolvers, overrides);
}
