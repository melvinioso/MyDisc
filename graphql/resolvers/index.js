import { merge } from 'lodash';
import GraphQLJSON from 'graphql-type-json';
import { GraphQLDateTime } from 'graphql-iso-date';

import Disc from './disc';

const scalars = {
  JSON: GraphQLJSON,
  DateTime: GraphQLDateTime,
};

const resolvers = merge(scalars, Disc);

export default resolvers;
