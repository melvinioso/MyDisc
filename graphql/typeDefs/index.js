import { gql } from 'apollo-server-express';

import Disc from './disc';

const Query = gql`
  scalar JSON
  scalar DateTime

  directive @listArgs on FIELD_DEFINITION
  directive @defaultArgs on FIELD_DEFINITION

  type Query {
    _empty: String
  }
`;

const Mutation = gql`
  type Mutation {
    _empty: String
  }
`;

const Subscription = gql`
  type Subscription {
    _empty: String
  }
`;

const typeDefs = [Query, Mutation, Subscription, Disc];

export default typeDefs;
