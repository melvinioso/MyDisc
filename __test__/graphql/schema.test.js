import { printSchema } from 'graphql';
import schema from '../../graphql/schema';

describe('GraphQL - Schema', () => {
  test('should match snapshot', async () => {
    expect(printSchema(schema)).toMatchSnapshot();
  });
});
