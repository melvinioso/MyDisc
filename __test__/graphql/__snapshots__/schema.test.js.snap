// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQL - Schema should match snapshot 1`] = `
"directive @listArgs on FIELD_DEFINITION

directive @defaultArgs on FIELD_DEFINITION

\\"\\"\\"
The \`JSON\` scalar type represents JSON values as specified by [ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf).
\\"\\"\\"
scalar JSON @specifiedBy(url: \\"http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf\\")

\\"\\"\\"
A date-time string at UTC, such as 2007-12-03T10:15:30Z, compliant with the \`date-time\` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.
\\"\\"\\"
scalar DateTime

type Query {
  _empty: String
  discs(order: String, where: JSON, offset: Int, limit: Int): [Disc]
  disc(id: Int!, where: JSON): Disc
}

type Mutation {
  _empty: String
  createDisc(disc: DiscCreate): Disc
  updateDisc(disc: DiscUpdate): Disc
  destroyDisc(disc: DiscDestroy): Disc
}

type Subscription {
  _empty: String
}

type Disc {
  id: Int
  brand: String
  mold: String
  plastic: String
  weight: Int
  speed: Int
  glide: Int
  turn: Int
  fade: Int
  createdAt: DateTime
  updatedAt: DateTime
}

input DiscCreate {
  brand: String
  mold: String
  plastic: String
  weight: Int
  speed: Int
  glide: Int
  turn: Int
  fade: Int
}

input DiscUpdate {
  id: Int!
  brand: String
  mold: String
  plastic: String
  weight: Int
  speed: Int
  glide: Int
  turn: Int
  fade: Int
}

input DiscDestroy {
  id: Int!
}
"
`;
